dartagnan/src/main/java/com/dat3m/dartagnan/verification/model/ValueModel.java

```diff
@@ -0,0 +1,43 @@
+package com.dat3m.dartagnan.verification.model;
+
+import java.math.BigInteger;
+
+
+// This class currently only serves as a representation of values in ExecutionModelNext,
+// in order to avoid the type conversion at value extraction.
+public class ValueModel {
+    private final Object value;
+    private final Type type;
+
+    public enum Type {
+        INTEGER, BOOLEAN
+    }
```

Comment:
This seems useless. Neither `type` is used in this class, nor `getType` from outside

Reply:
Yes. This `Type` should not be an enum but our existing `Type` class I believe.
Eventually we probably need a whole hierarchy of `ValueModels` like `IntegerModel`, `BooleanModel`, `PointerModel` etc. and this is why I thought that these various models are basically our literals: `IntLiteral`, `BoolLiteral` etc.
Using literals would certainly work for now, but might fail to work in the future (though depends on how we define literals for more complex types like pointers).

Reply:
OK. I will delete it. Basically the only meaning of existence for this rough class is to avoid type conversion at value extraction and to show all types of the values we have so far (int, bool and probably null) properly in witness. There should definitely be a real `ValueModel` that can represent all types of values and make `CAAT` running well.

Reply:
The class does indeed make the code look better. But with the current usage, I don't think there is much value in having the types there. If we have them, we should be using our real types rather than an enum as Thomas said.

Reply:
>even addresses are ValueModel and not BigInteger then

Why is that? Is it possible that representation for addresses could be other types than `BigInteger`?

Reply:
Yes. I mean the reason why we need `ValueModel` is we have different types of values, and we might need to store more information than literal (like size for example) about the value. But do we need this or should we use this to model memory address?

Reply:
Yes, addresses are also values. Eventually we will have pointers with provenance, which are tuples `(base, offset)`, and then you can also put address spaces on top and get `(addrSpace, base, offset)` pointers.
Then there are also pointers with capabilities in CHERI systems, and those are even larger than 64 bits.
I mean, it is easy to imagine pointers being more than an integer.

The general way to accommodate for such extensions is to have a correspondence between `Type` and `Value`, i.e., for every `Type` there is a class/object that represents concrete values of this `Type`.
Currently, `IntLiteral` and `BoolLiteral` fully satisfy this property and so would we good candidates.
However, I'm not so sure that this will be the case in the future (though maybe this is exactly what a `Literal` should be?). Anyway, I'm not saying you need to use the `Literal` classes now, but you should still use your `ValueModel` to represent addresses even if they happen to be plain integers now.

Reply:
I see. Then I think the `address` of `MemoryObjectModel` should be changed also, right?